### PR TITLE
[Feature] Create Raw Material Request from Sales Order

### PR DESCRIPTION
**Pull request by shreyashah115**
Closes: #15429

screen shot 2018-09-20 at 3 50 25 pm
screen shot 2018-09-20 at 3 50 09 pm
screen shot 2018-09-20 at 3 50 00 pm

bom-req

### CODE REVIEW

**Comment by rohitwaghchaure** [comment]
45815756-411a8080-bcf7-11e8-8a1a-01e6727491c8

move check box Include exploded items, ignore existing ordered qty below the for warehouse

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 1]
```diff
@@ -313,6 +315,71 @@ erpnext.selling.SalesOrderController = erpnext.selling.SellingController.extend(
...
+	make_raw_material_request: function() {
+			var me = this;
+			this.frm.call({
+				doc: this.frm.doc,
+				method: 'get_work_order_items',
+				args: {
+					for_raw_material_request: 1
+				},
+				callback: function(r) {
+					var fields = [
```
Comment (rohitwaghchaure) [inline erpnext/selling/doctype/sales_order/sales_order.js]: Add this part in separate method for readability

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 2]
```diff
@@ -520,3 +487,63 @@ def get_bin_details(row):
...
+@frappe.whitelist()
+def get_items_for_material_requests(doc, company=None):
+	if isinstance(doc, string_types):
+		doc = frappe._dict(json.loads(doc))
+
+	doc['mr_items'] = []
+	if doc.get('po_items'):
+		po_items = doc['po_items']
+	else:
+		po_items = doc['items']
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]: po_items = doc['po_items'] if doc.get('po_items') else doc['items']

**Comment by shreyashah115** [comment]
@rohitwaghchaure updated.

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 3]
```diff
@@ -520,3 +487,60 @@ def get_bin_details(row):
 	""".format(conditions=conditions), { "item_code": row.item_code }, as_list=1)
 
 	return item_projected_qty and item_projected_qty[0] or (0,0)
+
+@frappe.whitelist()
+def get_items_for_material_requests(doc, company=None):
```
Comment (rmehta) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]: always explicitly check permissions for whitelisted methods.

**Review by rmehta — DISMISSED** [review]

File: erpnext/manufacturing/doctype/production_plan/production_plan.py (hunk 3 above)
Comment (nabinhait) [reply erpnext/manufacturing/doctype/production_plan/production_plan.py]: @shreyashah115 Please fix this.

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 4]
```diff
@@ -520,3 +487,60 @@ def get_bin_details(row):
...
+			planned_qty = data.get('planned_qty')
+			bom_no = data.get('bom_no')
+			include_subcontracted_items = doc['include_subcontracted_items']
+			company = doc['company']
+			include_non_stock_items = doc['include_non_stock_items']
+			ignore_existing_ordered_qty = doc['ignore_existing_ordered_qty']
+		if not planned_qty:
+			frappe.throw(_("For row {0}: Enter Planned Qty").format(data.get('idx')))
+
+		if data.get('include_exploded_items') and bom_no and include_subcontracted_items:
```
Comment (nabinhait) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]: Can you break into multiple functions?
https://github.com/frappe/erpnext/wiki/Coding-Standards#break-functions--methods-more-than-10-lines-of-code

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 5]
```diff
@@ -520,3 +487,60 @@ def get_bin_details(row):
...
+			for d in frappe.db.sql("""select bei.item_code, item.default_bom as bom,
```
Comment (nabinhait) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]: Add comments why you needed this query. I think this query already part of the called function.

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 6]
```diff
@@ -367,15 +369,20 @@ def get_work_order_items(self):
 				bom = get_default_bom_item(i.item_code)
 				if bom:
 					stock_qty = i.qty if i.doctype == 'Packed Item' else i.stock_qty
-					pending_qty= stock_qty - flt(frappe.db.sql('''select sum(qty) from `tabWork Order`
-							where production_item=%s and sales_order=%s and sales_order_item = %s and docstatus<2''', (i.item_code, self.name, i.name))[0][0])
+					if not for_raw_material_request:
+						pending_qty = stock_qty - flt(frappe.db.sql('''select sum(qty) from `tabWork Order`
+								where production_item=%s and sales_order=%s and sales_order_item = %s and docstatus<2''', (i.item_code, self.name, i.name))[0][0])
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: use a variable to get `total_work_order_qty`.

**Review by nabinhait — DISMISSED** [review]

**Comment by nabinhait** [comment]
Fix conflict.

**Pull request merged**